changes/changes.go

```diff
@@ -0,0 +1,612 @@
+				log.Debugf("mean was nil")
+				return nil, qerr.New(ErrCompareStats, "nil stats values")
+			}
+			deltaCol["mean"] = rightStatsMap["mean"].(float64)
+		} else if rCol.Type.HasType("string") {
+			deltaCol["count"] = rightStatsMap["count"].(float64)
+			deltaCol["maxLength"] = rightStatsMap["maxLength"].(float64)
+			deltaCol["minLength"] = rightStatsMap["minLength"].(float64)
+			// TODO(arqu): investigate further - the bellow stats can be null
+			// and silently fail the request if the stats accumulator was not closed propperly
+			if rightStatsMap["unique"] == nil {
+				log.Debugf("unique was nil")
+				return nil, qerr.New(ErrCompareStats, "nil stats values")
+			}
+			deltaCol["unique"] = rightStatsMap["unique"].(float64)
+		} else if rCol.Type.HasType("boolean") {
+			deltaCol["count"] = rightStatsMap["count"].(float64)
+			deltaCol["trueCount"] = rightStatsMap["trueCount"].(float64)
+			deltaCol["falseCount"] = rightStatsMap["falseCount"].(float64)
+		}
+	} else {
+		return nil, fmt.Errorf("no left or right column present")
+	}
+
+	return deltaCol, nil
+}
+
+// Report computes the change report of two sources
+// This takes some assumptions - we work only with tabular data, with header rows and functional structure.json
+func (svc *Service) Report(ctx context.Context, leftRef, rightRef dsref.Ref, loadSource string) (*ChangeReportResponse, error) {
+	leftDs, err := svc.loader.LoadDataset(ctx, leftRef, loadSource)
+	if err != nil {
+		return nil, err
+	}
+	if rightRef.Path == "" {
+		rightRef.Path = leftDs.PreviousPath
+	}
+	rightDs, err := svc.loader.LoadDataset(ctx, rightRef, loadSource)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &ChangeReportResponse{}
+
+	leftVi := dsref.ConvertDatasetToVersionInfo(leftDs)
+	rightVi := dsref.ConvertDatasetToVersionInfo(rightDs)
+
+	res.VersionInfo = &ChangeReportComponent{}
+	res.VersionInfo.Left = leftVi
+	res.VersionInfo.Right = rightVi
+	res.VersionInfo.Meta = EmptyObject{}
+
+	if leftVi.Path == rightVi.Path {
+		res.VersionInfo.Meta["status"] = fsi.STUnmodified
+	} else {
+		res.VersionInfo.Meta["status"] = fsi.STChange
+	}
+
+	if leftDs.Commit != nil || rightDs.Commit != nil {
+		res.Commit = &ChangeReportComponent{}
+		if leftDs.Commit != nil {
+			res.Commit.Left = leftDs.Commit
+		} else {
+			res.Commit.Left = EmptyObject{}
+		}
+		if rightDs.Commit != nil {
+			res.Commit.Right = rightDs.Commit
+		} else {
+			res.Commit.Right = EmptyObject{}
+		}
+		res.Commit.Meta = EmptyObject{}
+
+		if leftDs.Commit != nil && rightDs.Commit == nil {
+			res.Commit.Meta["status"] = fsi.STRemoved
+		} else if leftDs.Commit == nil && rightDs.Commit != nil {
+			res.Commit.Meta["status"] = fsi.STAdd
+		} else if leftDs.Commit != nil && rightDs.Commit != nil {
+			if leftDs.Commit.Path == rightDs.Commit.Path {
+				res.Commit.Meta["status"] = fsi.STUnmodified
+			} else {
+				res.Commit.Meta["status"] = fsi.STChange
+			}
+		} else {
+			res.Commit.Meta["status"] = fsi.STParseError
```

Comment:
Are we sure this case means some file has failed to parse? Couldn't it also be both left and right have a nil component? Maybe it's better to have another status code in `fsi` such as `fsi.STMissing`.

Reply:
True, I just used a blanket error and tried to conform to what exists already due to existing integrations, happy to update.